### PR TITLE
Add some commands and fix graphics pipeline creation

### LINKED ISSUE #10: Wrong offsets in bind_vertex_buffers
https://github.com/MaikKlein/ash/blob/master/src/device.rs#L347

Offsets should be a slice not a reference

### PR DESCRIPTION
**Pull request by msiglreith**
- Add `clear_color_image` command
- Add `dispatch` command
- Fix vertex buffer binding, fixes #10 
- Fix graphics pipeline creation for multiple pipelines. (Return pipelines *and* error on failure to recover the correctly built pipelines)

### CODE REVIEW

**Comment by MaikKlein** [comment]
Looks good, thanks for the PR. It seems that the examples need to be updated. I can do that later today.

File: src/device.rs [hunk 1]
```diff
@@ -343,13 +357,15 @@ pub trait DeviceV1_0 {
...
     unsafe fn cmd_bind_vertex_buffers(&self,
                                       command_buffer: vk::CommandBuffer,
+                                      first_binding: vk::uint32_t,
                                       buffers: &[vk::Buffer],
-                                      offsets: &vk::DeviceSize) {
+                                      offsets: &[vk::DeviceSize]) {
+        debug_assert_eq!(buffers.len(), offsets.len());
         self.fp_v1_0().cmd_bind_vertex_buffers(command_buffer,
-                                               0,
+                                               first_binding,
```
Comment (MaikKlein) [inline src/device.rs]: Is this necessary? We already use a slice here which already has the information where it should start. For example

`device.cmd_bind_vertex_buffers(cb, &buffers[5 .. 10], &offsets);`

File: src/device.rs [hunk 2]
```diff
@@ -409,8 +436,8 @@ pub trait DeviceV1_0 {
                                        pipelines.as_mut_ptr());
         pipelines.set_len(create_infos.len());
         match err_code {
-            vk::Result::Success => Ok(pipelines),
-            _ => Err(err_code),
+            vk::Result::Success => (pipelines, Ok(())),
+            _ => (pipelines, Err(err_code)),
```
Comment (MaikKlein) [inline src/device.rs]: I am a bit confused after reading https://www.khronos.org/registry/vulkan/specs/1.0/man/html/vkCreateGraphicsPipelines.html

It seems that the function can only fail if the host or the device is oom. How can you recover pipelines that failed to build? I assumed that no pipeline would be returned if an error occurred.

File: src/device.rs (hunk 1 above)
Comment (msiglreith) [inline src/device.rs]: This should be necessary imo if you only want to update a part of the vertex input bindings,
`first_binding` states the first vertex input binding offset, else we would have to set it to 0 and have to update all input bindings.

Corresponding specs: (https://www.khronos.org/registry/vulkan/specs/1.0/man/html/vkCmdBindVertexBuffers.html)

> The values taken from elements i of pBuffers and pOffsets replace the current state for the vertex input binding firstBinding + i, for i in [0, bindingCount).

File: src/device.rs (hunk 2 above)
Comment (msiglreith) [inline src/device.rs]: Good point there. After reading the linked page I would agree with you.
I mostly followed the `Multiple Pipeline Creation` section.
> If any pipeline fails creation (for example, due
to out of memory errors), the vkCreate*Pipelines commands will return an error code. The
implementation will attempt to create all pipelines, and only return VK_NULL_HANDLE values for
those that actually failed.

The specification isn't really clear here I think.

Therefore, I tried a bit around to get one of these mysterious `VK_NULL_HANDLES`. Setting `s_type` in the `GraphicsPipelineCreateInfo` to `vk::StructureType::ComputePipelineCreateInfo` results in a `Err(ErrorInitializationFailed)` on my AMD system, which isn't listed as a return code!
Creating two pipelines (one of them failed) gave me this as return code:
`([0x0, 0x1edbeea5830], Err(ErrorInitializationFailed))`
So one null handle, a valid pipeline handle and an error.

I'm not sure if this is correct behavior from the driver or the specs are missing something, but imo the proposed code change handles this scenario more gracefully.

**Comment by msiglreith** [comment]
Sorry, for missing to update the examples. I can also take care of it if you want (:

**Comment by MaikKlein** [comment]
I already did it, but I didn't push it yet. I'll push it tomorrow.

File: src/device.rs (hunk 2 above)
Comment (MaikKlein) [inline src/device.rs]: I see, I guess that makes sense.

File: src/device.rs (hunk 1 above)
Comment (MaikKlein) [inline src/device.rs]: You are right, I actually misunderstood the purpose of firstBinding.

File: src/device.rs (hunk 2 above)
Comment (MaikKlein) [inline src/device.rs]: I think i might slightly adjust the return type to `Result<Vec<vk::Pipeline>, (Vec<vk::Pipeline>, vk::Result)>`.  This guarantees that on success that there will not be any null handles inside the vec, and you can still recover if you need too. Any objections?

Comment (msiglreith) [inline src/device.rs]: Sounds good! 👍

**Pull request merged**